### PR TITLE
perf(functions): Speed up date_trunc/trunc on DATE and TIMESTAMP via Neri-Schneider

### LINKED ISSUE #17420: perf: Speed up date_trunc / trunc on DATE and TIMESTAMP using Neri-Schneider
### Description

Follow-up to #17371 (which closed #17364 for the date extraction direction).

Both Presto's `date_trunc(unit, ...)` and Spark's `trunc(date, format)` and `date_trunc(unit, ...)` route through the same `getDateTime` + `adjustDateTime` + `Timestamp::calendarUtcToEpoch` chain in `velox/functions/lib/TimeUtils.cpp` for date-level truncation units (`kWeek`, `kMonth`, `kQuarter`, `kYear`):

1. multiply by 86400 to get seconds (DATE input) or use `getSeconds(timestamp, timeZone)` (TIMESTAMP input)
2. fill a full `std::tm` via `Timestamp::epochToCalendarUtc` (`tm_year/mon/mday/yday/wday/hour/min/sec/isdst`)
3. mutate it to "start of unit" via `adjustDateTime`
4. convert back via `Timestamp::calendarUtcToEpoch` (loops through `daysBetweenYears` and the leap-month-offset table)

For date-level units the entire `std::tm` round trip is unnecessary - the answer can be expressed directly on `(year, month, day)` (kYear/kMonth/kQuarter) or pure integer arithmetic on the day count itself (kWeek). The Neri-Schneider primitives from #17371 (`daysToYmd`, `ymdToDays`) plus a small day-of-week computation are enough.

### Proposed approach

Add a `truncateDate(days, unit)` helper in `velox/functions/lib/TimeUtils`:

- `kDay`: identity.
- `kWeek`: pure integer math. 1970-01-01 was a Thursday, so `(4 + days) mod 7` gives the ISO weekday; shift back to Monday with one branch.
- `kMonth/kQuarter/kYear`: `daysToYmd(days)` → adjust month → `ymdToDays(year, adjustedMonth, 1)`. Falls back to the existing `std::tm` path only at the boundary years where `ymdToDays` is unsafe.

Wire it in three places:

- Presto `DateTruncFunction::call(Date, ...)` (`velox/functions/prestosql/DateTimeFunctions.h`)
- Spark `TruncFunction::call(Date, ...)` (`velox/functions/sparksql/DateTimeFunctions.h`)
- The `default` branch of `truncateTimestamp` in `TimeUtils.cpp` (which handles `kWeek`/`kMonth`/`kQuarter`/`kYear` for `TIMESTAMP` inputs). The TIMESTAMP variant floors local seconds to days, calls `truncateDate`, then re-materializes a local `Timestamp` at midnight; the existing `toGMT` round trip below preserves timezone semantics.

Sub-day units (`kMicrosecond`/`kMillisecond`/`kSecond`/`kMinute`/`kHour`) and the timezone-handling already-fast `kDay` branch are unchanged. The boundary-year fallback path is also unchanged so observable behavior is preserved across the full day-count range.

### PR DESCRIPTION
**Pull request by Licht-T**
## Summary
Adds a `truncateDate(days, unit)` helper in `velox/functions/lib/TimeUtils` that bypasses the `std::tm` round trip used by `getDateTime + adjustDateTime + calendarUtcToEpoch`:

- `kDay` returns `days` unchanged.
- `kWeek` shifts back to the most recent Monday with pure integer math on `(4 + days) mod 7` (1970-01-01 was a Thursday).
- `kMonth/kQuarter/kYear` go directly through Neri-Schneider - `daysToYmd` → adjust month → `ymdToDays`.
- Falls back to the existing `std::tm` path only at the rare boundary years where `ymdToDays` is unsafe.

Wires it into three call sites:
- Presto `DateTruncFunction::call(Date, ...)`
- Spark `TruncFunction::call(Date, ...)`
- The `default` branch of `truncateTimestamp` (week/month/quarter/year on TIMESTAMP), which floors local seconds to days, calls `truncateDate`, then re-materializes a local `Timestamp` at midnight; the existing `toGMT` round trip below preserves timezone semantics. Falls back to the prior `std::tm` path only when `localDays` overflows int32 (essentially unreachable for any practical Velox `Timestamp`).

Sub-day truncation units (kMicrosecond/kMillisecond/kSecond/kMinute/kHour) and the already-fast kDay branch are unchanged.

Closes #17420. Follow-up to #17371 / #17364.

## Numbers

`velox_date_extract_benchmark --bm_regex=date_trunc_` (1024-row vectors, bounded ranges so the measurement reflects in-range work, 3-run median, GCC 11 release):

DATE input:
| unit | baseline | patched | speedup |
|------|---------:|--------:|--------:|
| year    | 11.68 ms | 4.00 ms | **2.92×** |
| quarter | 12.90 ms | 5.58 ms | **2.31×** |
| month   | 12.05 ms | 4.87 ms | **2.47×** |
| week    | 12.30 ms | 3.04 ms | **4.05×** |

TIMESTAMP input:
| unit | baseline | patched | speedup |
|------|---------:|--------:|--------:|
| year    | 22.09 ms | 13.83 ms | **1.60×** |
| quarter | 24.09 ms | 14.67 ms | **1.64×** |
| month   | 22.46 ms | 14.48 ms | **1.55×** |
| week    | 23.78 ms | 11.73 ms | **2.03×** |

The `kWeek` path is purest int math (no `daysToYmd`/`ymdToDays`), which is why it wins biggest. The TIMESTAMP speedups are smaller because the per-row constant overhead (timezone bookkeeping, Timestamp construction, optional `toGMT`) is unchanged - only the std::tm round trip is replaced.

## Correctness

- All Presto `DateTimeFunctionsTest.dateTrunc*` tests pass (13 tests, including timestamp-with-timezone variants like `dateTruncTimeStampForWeek`, `dateTruncTimeStampWithTimezoneForWeek`, `dateTruncTimeStampWithTimezoneStringForWeek`).
- All Spark `DateTimeFunctionsTest.dateTrunc / trunc / truncate*` tests pass (4 tests).
- The boundary-year fallback path retains the `getDateTime + adjustDateTime + calendarUtcToEpoch` flow, so any observable behavior change for inputs outside the Neri-Schneider domain would also surface a regression in the existing test suite - none observed.

## Test

- `make unittest` on a release build: `velox_functions_test` (Presto DateTimeFunctionsTest), `velox_functions_spark_test` (Spark DateTimeFunctionsTest).
- `velox_date_extract_benchmark --bm_regex=date_trunc_` to validate per-unit speedups on the reviewer's hardware.

### CODE REVIEW

**Comment by Licht-T** [comment]
@kgpai PTAL

**Comment by kgpai** [comment]
Hi @Licht-T  , Thank you for this valuable contribution. Please give me a day or two to get to this.

**Comment by kgpai** [comment]
Could you add the following to `velox/functions/lib/tests/TimeUtilsTest.cpp`? The new integer-arithmetic path doesn't have direct unit coverage today, and the existing `dateTrunc*` suites don't exercise the boundaries, fallback path, or DST-spanning truncations.

**1. Direct `truncateDate` coverage:**
```
truncateDate(0,  kDay)     == 0
truncateDate(0,  kWeek)    == -3   // 1970-01-01 was Thursday → Mon = 1969-12-29
truncateDate(0,  kMonth)   == 0
truncateDate(0,  kYear)    == 0
truncateDate(0,  kQuarter) == 0
truncateDate(4,  kWeek)    == 4    // Monday
truncateDate(3,  kWeek)    == -3   // Sunday → previous Monday
```

**2. Negative `days` / pre-epoch:**
```
truncateDate(-1, kDay)     == -1   // 1969-12-31
truncateDate(-1, kWeek)    == -3
truncateDate(-1, kMonth)   == -31  // 1969-12-01
truncateDate(-1, kYear)    == -365
truncateDate(-7, kWeek)    == -10
```

**3. INT32 boundaries for `kWeek`** — `(4 + INT32_MAX) % 7` is signed overflow:
```
truncateDate(std::numeric_limits<int32_t>::max(), kWeek)
truncateDate(std::numeric_limits<int32_t>::min(), kWeek)
```

**4. FastDate fallback path** — currently uncovered:
```
truncateDate(daysFor(kYearMax - 1, 6, 15), kYear)
truncateDate(daysFor(kYearMin + 1, 6, 15), kMonth)
EXPECT_THROW(truncateDate(std::numeric_limits<int32_t>::max(), kYear), VeloxUserError);
```

**5. `truncateTimestamp` across DST transitions:**
```cpp
auto la = locateZone("America/Los_Angeles");
// 2026-03-15 17:00 UTC = 2026-03-15 10:00 PDT.
// kWeek → Mon 2026-03-09 00:00 PDT = 2026-03-09 07:00 UTC.
EXPECT_EQ(
    truncateTimestamp(Timestamp(1773450000, 0), DateTimeUnit::kWeek, la),
    Timestamp(1772614800, 0));

// 2026-02-15 18:00 UTC = 2026-02-15 10:00 PST.
// kMonth → 2026-02-01 00:00 PST = 2026-02-01 08:00 UTC.
```

**6. Pre-epoch + negative-offset tz** — exercises floor-division end-to-end:
```cpp
// 1969-12-31 23:30 UTC = 1969-12-31 15:30 PST.
// kWeek → Mon 1969-12-29 00:00 PST = 1969-12-29 08:00 UTC.
truncateTimestamp(Timestamp(-1800, 0), DateTimeUnit::kWeek, la)
```

**7. Non-integer-hour offsets** — at least one kWeek and one kMonth case in `Asia/Kolkata` (+5:30) and `Asia/Kathmandu` (+5:45). Not exercised today.

File: velox/functions/lib/TimeUtils.cpp [hunk 1]
```diff
@@ -188,16 +189,96 @@ Timestamp truncateTimestamp(
...
+int32_t truncateDate(int32_t days, DateTimeUnit unit) {
+  if (unit == DateTimeUnit::kDay) {
+    return days;
+  }
+
+  if (unit == DateTimeUnit::kWeek) {
+    // 1970-01-01 was a Thursday, so for any day d the ISO weekday
+    // (0 = Sun, 1 = Mon, ..., 6 = Sat) is `(4 + d) mod 7`. To shift
+    // back to Monday: Sun -> 6 days back, Mon -> 0, otherwise
+    // weekday - 1.
```
Comment (mbasmanova) [inline velox/functions/lib/TimeUtils.cpp:230]: `(4 + days)` overflows `int32_t` when `days == INT32_MAX`. UB per the standard; sanitizers will catch it.

Suggest: `int32_t weekday = static_cast<int32_t>((4LL + days) % 7);`

File: velox/functions/lib/TimeUtils.cpp [hunk 2]
```diff
@@ -188,16 +189,96 @@ Timestamp truncateTimestamp(
...
-      result = Timestamp(Timestamp::calendarUtcToEpoch(dateTime), 0);
+    default: {
+      // Week, month, quarter, and year truncation are date-unit
+      // operations: convert to local days, truncate via the
+      // Date-domain helper (which avoids the std::tm round trip and
+      // goes directly through Neri-Schneider for in-range years),
+      // then materialize a local Timestamp at the start of the
+      // truncated unit. The toGMT below converts it back. Falls back
+      // to the std::tm path for the unreachable extreme inputs whose
+      // local day count overflows int32_t.
```
Comment (mbasmanova) [inline velox/functions/lib/TimeUtils.cpp]: This 7-line comment is longer than the code it documents. `truncateDate` is descriptively named; a one-liner would suffice, e.g.

```cpp
// Convert to local days, truncate, convert back to local midnight.
```

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@Licht-T, Thank you for the contribution. Nice speedups!

A few comments.

## Bug: signed integer overflow in `kWeek`

`TimeUtils.cpp` — `truncateDate`, the week branch:

```cpp
int32_t weekday = (4 + days) % 7;
```

`4 + days` overflows `int32_t` when `days == INT32_MAX` (2,147,483,651 doesn't fit). This is undefined behavior. Sanitizers will flag it.

Suggest:

```cpp
int32_t weekday = static_cast<int32_t>((4LL + days) % 7);
```

## Comments

The 7-line comment in the `truncateTimestamp` default branch is longer than the code it documents. `truncateDate` is descriptively named; a one-liner would suffice, e.g.

```cpp
// Convert to local days, truncate, convert back to local midnight.
```

## Tests

Please, add unit tests for `truncateDate` to `velox/functions/lib/tests/TimeUtilsTest.cpp` along the lines @kgpai suggested.

**Review by rui-mo — COMMENTED** [review]
Thanks for this optimization. Since this change introduces a fast path with a fallback path, could we add boundary regression tests? In particular: negative timestamps, values near the kRataDieMin/kRataDieMax limits, and DST boundary days. These cases would help verify that fast-path and fallback-path results remain consistent.

File: velox/functions/lib/TimeUtils.cpp (hunk 2 above)
Comment (Licht-T) [reply velox/functions/lib/TimeUtils.cpp]: Fixed.

File: velox/functions/lib/TimeUtils.cpp (hunk 1 above)
Comment (Licht-T) [reply velox/functions/lib/TimeUtils.cpp:230]: Good catch, fixed. While adding the boundary tests you suggested I found the *same* branch overflows again one line down: `days - offsetBack` underflows `int32_t` when `days == INT32_MIN`. So I hardened both:

```cpp
int32_t weekday = static_cast<int32_t>((4LL + days) % 7);
if (weekday < 0) {
  weekday += 7;
}
const int32_t offsetBack = weekday == 0 ? 6 : weekday - 1;
// Subtract in 64-bit to avoid signed overflow when days == INT32_MIN.
return static_cast<int32_t>(static_cast<int64_t>(days) - offsetBack);
```

**Comment by Licht-T** [comment]
Thanks for the insightful reviews!

@kgpai This was a great checklist - added all of it to `TimeUtilsTest.cpp` as two tests (`truncateDate` and `truncateTimestampTimeZone`).

Two corrections worth flagging, since the final values differ from the snippet:

1. **The DST epoch literals were slightly off.** `2026-03-15 17:00 UTC` is epoch `1773594000` (the quoted `1773450000` is ~1.7 days earlier).
2. **`truncateDate(INT32_MAX, kYear)` does not throw.** It's outside the FastDate range, so `epochToCalendarUtc` routes it through `WideRangeDateConversion`, which handles the whole `int32` day range. I replaced the `EXPECT_THROW` with exact-value assertions on both sides of the `kRataDie` boundary instead.

@rui-mo The new `truncateDate` test asserts results just inside the FastDate range (fast path, near `kYearMin`/`kYearMax`) and just outside it (`kRataDieMax + 1`, `kRataDieMin - 1`, `INT32_MAX`)

File: velox/functions/lib/TimeUtils.cpp [hunk 3]
```diff
@@ -188,16 +189,90 @@ Timestamp truncateTimestamp(
...
+    // (0 = Sun, 1 = Mon, ..., 6 = Sat) is `(4 + d) mod 7`. To shift
+    // back to Monday: Sun -> 6 days back, Mon -> 0, otherwise
+    // weekday - 1.
+    int32_t weekday = static_cast<int32_t>((4LL + days) % 7);
+    if (weekday < 0) {
+      weekday += 7;
+    }
+    const int32_t offsetBack = weekday == 0 ? 6 : weekday - 1;
+    // Subtract in 64-bit to avoid signed overflow when days == INT32_MIN.
+    return static_cast<int32_t>(static_cast<int64_t>(days) - offsetBack);
```
Comment (rui-mo) [inline velox/functions/lib/TimeUtils.cpp]: Could casting the value back to int32_t produce unexpected results when the `days` is INT32_MIN and `offsetBack` is positive?

**Review by rui-mo — COMMENTED** [review]
Thanks. I left one more comment.

File: velox/functions/lib/TimeUtils.cpp (hunk 3 above)
Comment (Licht-T) [reply velox/functions/lib/TimeUtils.cpp]: Yes, and this is exactly same behavior what the pre-PR `std::tm` path did too (confirmed sweeping each other). But, silently returning a wrapped date isn't good, so rather than rely on that I've added an explicit check:

```cpp
// week
const int64_t weekStart = static_cast<int64_t>(days) - offsetBack;
VELOX_USER_CHECK_GE(
    weekStart, std::numeric_limits<int32_t>::min(),
    "Date is out of range for truncation: {} days", days);
return static_cast<int32_t>(weekStart);

// std::tm fallback (month / quarter / year)
const int64_t truncatedDays =
    Timestamp::calendarUtcToEpoch(dateTime) / kSecondsInDay;
VELOX_USER_CHECK_GE(
    truncatedDays, std::numeric_limits<int32_t>::min(),
    "Date is out of range for truncation: {} days", days);
return static_cast<int32_t>(truncatedDays);
```

**Review by rui-mo — APPROVED** [review]
